packages/vm/src/context.rs

```diff
@@ -196,8 +196,21 @@ pub fn get_gas_state<'a, 'b, S: Storage, Q: Querier + 'b>(ctx: &'a mut Ctx) -> &
     &mut get_context_data_mut::<S, Q>(ctx).gas_state
 }
 
+/// Use this function to adjust the VM's gas limit when a call into the backend
+/// reported there was externally metered gas used.
+/// This does not increase the VM's gas usage but ensures the overall limit is not exceeded.
+pub fn account_for_externally_used_gas<S: Storage, Q: Querier>(
```

Comment:
First, I like the new name - very clear.

Second, why do we need this to call the different impl? Can't we just have the original called function under feature gate? (Or is this to get better rust docs?)

Reply:
The idea is to have exactly one public function in all configurations that has one doc comment attached. Then the implementation can vary based on features. This is needed for RLS based IDEs where the language server might have a different config than your compilation.

But even better than swapping out *_impl functions is the possibility to swap out implementation scopes like this:

```rust
/// Use this function to adjust the VM's gas limit when a call into the backend
/// reported there was externally metered gas used.
/// This does not increase the VM's gas usage but ensures the overall limit is not exceeded.
#[allow(unused)]
pub fn account_for_externally_used_gas<S: Storage, Q: Querier>(
    ctx: &mut Ctx,
    amount: u64,
) -> VmResult<()> {
    #[cfg(feature = "default-singlepass")]
    {
        use crate::backends::{get_gas_left, set_gas_limit};

        let ctx_data = get_context_data_mut::<S, Q>(ctx);
        if let Some(mut instance_ptr) = ctx_data.wasmer_instance {
            // and so on …

            if gas_state.externally_used_gas + wasmer_used_gas > gas_state.gas_limit {
                Err(VmError::GasDepletion)
            } else {
                Ok(())
            }
        } else {
            Err(VmError::uninitialized_context_data("wasmer_instance"))
        }
    }

    #[cfg(feature = "default-cranelift")]
    {
        Ok(())
    }
}
```

Reply:
Looking at this alternative now, I'm not conviced it is better than multiple `account_for_externally_used_gas_impl` implementations

Reply:
Thanks for explaining.